web/src/hooks/useInfiniteQueryPosts.ts

```diff
@@ -0,0 +1,76 @@
+import { useInfiniteQuery } from "@tanstack/react-query";
+import { useScroll } from "framer-motion";
+import { useEffect, useMemo } from "react";
+import { toast } from "react-hot-toast";
+
+import { searchPosts } from "../requests/fetch";
+import useDebounce from "./useDebounce";
+
+interface Params {
+  key: string;
+  searchValue: string;
+  userId?: string;
+  recentOnly?: boolean;
+  limit?: number;
+  queryOptions?: {
+    refetchOnMount?: boolean | "always";
+    staleTime?: number;
+    enabled?: boolean;
+  };
+}
+
+export default function useInfiniteQueryPosts({
+  key,
+  searchValue,
+  userId,
+  limit,
+  recentOnly,
+  queryOptions,
+}: Params) {
+  const { scrollYProgress } = useScroll();
+
+  const debouncedSearchValue = useDebounce(searchValue, 250);
+
+  const {
+    data: infinitePostsData,
+    fetchNextPage,
+    hasNextPage,
+    isFetching,
+    isFetchingNextPage,
+  } = useInfiniteQuery({
+    queryKey: [key, { search: debouncedSearchValue, userId }],
+    queryFn: ({ pageParam = "" }) =>
+      searchPosts({
+        search: searchValue,
+        userId,
+        cursorId: pageParam as string,
+        limit,
+        recentOnly,
+      }),
+    getNextPageParam: (lastPage) => lastPage.nextCursorId,
+    onError: (err: Error) => {
+      toast.error(err.message);
+    },
+    refetchOnMount: queryOptions?.refetchOnMount ?? true,
+    staleTime: queryOptions?.staleTime ?? 1000 * 60 * 5,
+    enabled: queryOptions?.enabled ?? true,
+  });
+
+  const posts = useMemo(() => {
+    return infinitePostsData?.pages.map((page) => page.posts).flat() ?? [];
+  }, [infinitePostsData]);
+
+  useEffect(() => {
+    return scrollYProgress.onChange((progress) => {
+      if (progress > 0.6 && !isFetchingNextPage && hasNextPage) {
+        void fetchNextPage();
+      }
+    });
+  }, [hasNextPage, isFetchingNextPage, scrollYProgress]);
```

Comment:
Eslint says missing dependency and I think I agree. We have a number of `useEffect()`s that are a bit light on dependencies across the codebase. Including the dependencies at worst means an extra call or two, but not having them can lead to very subtle bugs. I tend to err in favor of overspecifying the dependencies because I'm not that smart.

Reply:
I dont fully understand what the purpose of including functions in the dependencies array is. I only need to rerun the logic inside the useEffect whenever those values change but not if the function reference changes

Reply:
So I can't say my understanding is complete myself. [This part of the docs](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) has me scared into just adding them by default.

One thing to note is that the dependencies array is really mostly an optimization. React doesn't make the guarantee that your code will only rerun if one of the dependencies change, only that if a dependency changes, the code _will_ rerun.

Reply:
If your effect relies on being called only when the dependencies change, then according to the React team it is buggy and should be rewritten.